src/transforms/kubernetes/watch_client.rs

```diff
@@ -0,0 +1,469 @@
+
+        let (parts, body) = request.into_parts();
+        Ok(Request::from_parts(parts, body.into()))
+    }
+
+    fn authorize(&self, request: &mut Request<Vec<u8>>) -> Result<(), BuildError> {
+        if let Some(token) = self.token.as_ref() {
+            request.headers_mut().insert(
+                header::AUTHORIZATION,
+                header::HeaderValue::from_str(format!("Bearer {}", token).as_str())
+                    .context(InvalidToken)?,
+            );
+        }
+
+        Ok(())
+    }
+
+    fn fill_uri(&self, request: &mut Request<Vec<u8>>) -> Result<(), BuildError> {
+        let mut uri = self.server.clone().into_parts();
+        uri.path_and_query = request.uri().clone().into_parts().path_and_query;
+        *request.uri_mut() = Uri::from_parts(uri).context(InvalidUriParts)?;
+        Ok(())
+    }
+}
+
+/// Kubernetes client for watching changes on Kubernetes Objects.
+pub struct WatchClient {
+    config: ClientConfig,
+    client: hyper::Client<HttpsConnector<HttpConnector<Resolver>>>,
+}
+
+impl WatchClient {
+    /// Builds Stream of newest Pod metadata.
+    /// With version None, will also stream inital Pod metadata.
+    ///
+    /// Caller should maintain latest `pod.metadata.resource_version` and must stop
+    /// using stream on first RuntimeError and start a new one.
+    ///
+    /// When encountering end of stream, caller should use RuntimeError::WatchUnexpectedlyEnded.
+    ///
+    /// Arguments:
+    ///  - `from` should contain latest `pod.metadata.resource_version`.
+    ///  - `error` must contain RuntimeError with which last stream ended.
+    pub fn watch_metadata(
+        &mut self,
+        mut version: Option<Version>,
+        error: Option<RuntimeError>,
+    ) -> Result<impl Stream<Item = Pod, Error = RuntimeError>, BuildError> {
+        match error {
+            None => (),
+            Some(RuntimeError::WatchEventError { status }) if status.code == Some(410) => {
+                // 410 Gone, restart with new list.
+                // https://kubernetes.io/docs/reference/using-api/api-concepts/#410-gone-responses
+                info!("Watch list desynced for Kubernetes Pod list. Restarting version.");
+                version = None;
+            }
+            Some(error) => debug!(%error),
+        }
+
+        // If watch is initiated with None resource_version, we will receive initial
+        // list of data as synthetic "Added" events.
+        // https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
+        let request = self.config.build_request(version)?;
+
+        Ok(self.build_watch_stream(request))
+    }
+
+    fn build_watch_stream(
+        &self,
+        request: Request<Body>,
+    ) -> impl Stream<Item = Pod, Error = RuntimeError> {
+        let mut decoder = Decoder::default();
+
+        self.client
+            .request(request)
+            .map_err(|error| RuntimeError::FailedConnecting { source: error })
```

Comment:
We could probably use https://docs.rs/snafu/0.6.2/snafu/futures01/trait.FutureExt.html here, what do you think?

Reply:
We could.